src/query_plan/generate.rs

```diff
@@ -0,0 +1,317 @@
+///   (and theoretically can be empty, though very very rarely is in practice).
+/// - `to_add` is the list of additional elements to add to `initial`
+///   to make a full plan of the query being planned.
+///   Each element of `to_add` corresponds to one of the query "leaf"
+///   and is itself a list of all the possible options for that "leaf".
+///
+/// In other words, a complete plan is obtained
+/// by picking one choice in each of the element of `to_add` (so picking `to_add.len()` element)
+/// and adding them all to `initial`.
+/// The question being, which particular choice for each element of `to_add` yield the best plan.
+///
+/// Of course, the total number of possible plans is the cartesian product of `to_add`,
+/// which can get large, and so this method is trying to trim some of the options.
+/// For that, the general idea is that we first generate one of the plan, compute its cost,
+/// and then as we build other options,
+/// we can check as we pick elements of `to_add` the cost of what we get,
+/// and if we ever get a higher cost than the one fo the complete plan we already have,
+/// then there is no point in checking the remaining elements,
+/// and we can thus cut all the options for the remaining elements.
+/// In other words, if a partial plan is ever already more costly
+/// than another full plan we have computed, then adding more will never get us a better plan.
+///
+/// Of course, this method is not guaranteed to save work,
+/// and in the worst case, we'll still generate all plans.
+/// But when a "good" plan is generated early, it can save a lot of computing work.
+///
+/// And the 2nd "trick" of this method is that it starts by generating the plans
+/// that correspond to picking choices in `to_add` at the same indexes,
+/// and this because this often actually generate good plans.
+/// The reason is that the order of choices for each element of `to_add` is not necessarily random,
+/// because the algorithm generating paths is not random either.
+/// In other words, elements at similar indexes have some good chance
+/// to correspond to similar choices, and so will tend to correspond to good plans.
+///
+/// Parameters:
+/// * `initial`: the initial partial plan to use.
+/// * `to_add`: a list of the remaining "elements" to add to `initial`.
+///   Each element of `to_add` correspond to multiple choice
+///   we can use to plan that particular element.
+/// * `add_function`: how to obtain a new plan by taking some plan and adding a new element to it.
+/// * `cost_function`: how to compute the cost of a plan.
+/// * `on_plan`: an optional method called on every _complete_ plan generated by this method,
+///   with both the cost of that plan and the best cost we have generated thus far
+///   (if that's not the first plan generated).
+///   This mostly exists to allow some debugging.
+pub fn generate_all_plans_and_find_best<Plan, Element>(
+    initial: Plan,
+    to_add: Vec<Choices<Element>>,
```

Comment:
One thing to note for `to_add`: its type here is `Vec<Vec<Option<T>>>`, but it doesn't really make sense to pass in a `None` here (that's a value used by this function to indicate a choice has been searched), so it would be nice to either:
1) Have it be `Vec<Vec<T>>` and convert it in the function, or
2) Update the docstring to indicate callers shouldn't pass in `None`.

Reply:
I’ve updated the doc comment, but either way a raw `Vec<Vec<_>>` (as opposed to `Vec` of some struct that contains `Vec`) feels a bit unusual. What do we expect the data structures to look like at the non-test call sites? Could we adapt the function so that no potentially-expensive conversion is needed?

Reply:
@SimonSapin 
In the JS codebase, the part that generates `to_add` looks like:
```typescript
const otherTrees = this
  .closedBranches
  .slice(0, idxFirstOfLengthOne)
  .map(b => b.map(opt => PathTree.createFromOpPaths(federatedQueryGraph, root, flattenClosedPath(opt))));
```
The map-within-a-map would presumably create the `Vec<Vec<_>>` (I imagine we could just put a `Some()` there to get the `Option<T>`, and leave a comment about why).